### PR TITLE
Net: Improve blocks only mode.

### PR DESCRIPTION
**Pull request by pstratem**
- Process inv messages in blocks only mode if we would relay the peers transactions
- Bail early in tx message processing if we will neither AcceptToMemoryPool nor RelayTransaction

### CODE REVIEW

**Comment by dcousens** [comment]
concept ACK, utACK

**Comment by paveljanik** [comment]
utACK

We should probably document/popularize blocks only mode, because it can be very useful for some usecases - thank you for it!

**Comment by tulip0** [comment]
@paveljanik 
When documenting some care needs to be taken to ensure people don't enable this on their nodes thinking it will let them receive blocks faster, "not processing transaction messages will make my node faster doing blocks" isn't an unreasonable assumption to make. In actuality running with `blocksonly`  has a pretty heavy impact on the speed of block verification by negating the ECDSA signature cache.

**Comment by gmaxwell** [comment]
The impact on mining is much of why I wasn't thinking to promote it yet.  We also yet don't do smart things with peers that send no relay.  E.g. a node that somehow ends up with 8 blocksonly peers won't do anything smart about it, and probably we'd like to do something about that before encouraging it's widespread use.

**Comment by dcousens** [comment]
@gmaxwell wasn't this going undocumented (or at least, a huge "highly discouraged" warning) next to if it at all made it in to `0.12`?

**Comment by pstratem** [comment]
The help message is only displayed in debug more.

File: src/main.cpp [hunk 1]
```diff
@@ -4243,6 +4246,13 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
                     LogPrint("net", "getheaders (%d) %s to peer=%d\n", pindexBestHeader->nHeight, inv.hash.ToString(), pfrom->id);
                 }
             }
+            else
+            {
+                if (fBlocksOnly)
+                    LogPrint("net", "peer sent inv %s in violation of protocol peer=%d\n", inv.ToString(), pfrom->id);
```
Comment (petertodd) [inline src/main.cpp]: Maybe I'm missing something, but why wouldn't a peer send us a inv? We're advertising NODE_NETWORK, and we haven't sent them a bloom filter.

Comment (pstratem) [inline src/main.cpp]: In the previous PR (which was merged into master) we set fRelayTxs = false in the version message in blocks only mode.

**Comment by petertodd** [comment]
@gmaxwell Re: the only blocksonly peers risk, it probably makes sense to have blocksonly unset NODE_NETWORK for now.

File: src/main.cpp (hunk 1 above)
Comment (petertodd) [inline src/main.cpp]: Weird, I can't seem to find the actual code that does that. :(

Comment (petertodd) [inline src/main.cpp]: Ah! Sorry, I'm blind, found it!

**Comment by gmaxwell** [comment]
@petertodd I'd rather not, we still serve and relay blocks. You can IBD off a blocks only node, etc.  There are already quite a few nodes out there that silently won't relay transactions, at least with the option undocumented I don't think it's a huge marginal risk.

File: src/main.cpp [hunk 2]
```diff
@@ -4367,6 +4377,14 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
     else if (strCommand == "tx")
     {
+        // Stop processing the transaction early if
+        // We are in blocks only mode and peer is either not whitelisted or whitelistalwaysrelay is off
+        if (GetBoolArg("-blocksonly", DEFAULT_BLOCKSONLY) && (!pfrom->fWhitelisted || !GetBoolArg("-whitelistalwaysrelay", DEFAULT_WHITELISTALWAYSRELAY)))
+        {
+            LogPrint("net", "peer sent transaction in violation of protocol peer=%d\n", pfrom->id);
```
Comment (petertodd) [inline src/main.cpp]: Reads better if reworded to say "peer %d sent"

File: src/main.cpp (hunk 1 above)
Comment (petertodd) [inline src/main.cpp]: Also reads better if reworded to say "peer %d sent"

**Comment by petertodd** [comment]
Tests done:

[x] Triggered inv and tx "violation of protocol" cases
[x] Tested with a bitcoinj wallet, which just ignored the fRelayTxs

ACK (modulo https://github.com/pstratem/bitcoin/pull/3)

**Comment by petertodd** [comment]
@gmaxwell Fair enough; I'm convinced.

**Comment by gmaxwell** [comment]
Sorry to nitpick: but I think these log entries should be made consistent:

2015-11-21 23:42:08 getblocks 384665 to 00000000000000000c8d710c468516e11f76188c18def174581317b8faac716d limit 500 from peer=86
2015-11-21 23:42:08   getblocks stopping at 384666 00000000000000000c8d710c468516e11f76188c18def174581317b8faac716d
2015-11-21 23:42:08 sending: inv (37 bytes) peer=86
2015-11-21 23:42:08 received: pong (8 bytes) peer=86
2015-11-21 23:42:08 sending: ping (8 bytes) peer=56
2015-11-21 23:42:08 received: tx (29246 bytes) peer=1
2015-11-21 23:42:08 peer 1 sent transaction in violation of protocol


E.g. "transaction (txid?) sent in violation of protocol peer=%d"

**Comment by gmaxwell** [comment]
@petertodd come reack that I undid your nit. :)

**Comment by petertodd** [comment]
@gmaxwell That wording looks fine, although I'd point out that phrasing it "[THING] sent in violation of protocol, peer=%d" - note the added comma! - should be fine too.

What can I call that? A μnit? :)

**Comment by gmaxwell** [comment]
ACK

**Pull request merged**